okapi/services/fieldnotes/upload/WebService.php

```diff
@@ -0,0 +1,175 @@
+                    ) values (
+                        '".Db::escape_string($user_id)."',
+                        '".Db::escape_string($geocache_id)."',
+                        '".Db::escape_string($type)."',
+                        '".Db::escape_string($date)."',
+                        '".Db::escape_string($text)."'
+                    )
+                ");
+
+            }
+            $result = array(
+                'success'          => true,
+                'totalRecords'     => $notes['totalRecords'],
+                'processedRecords' => $notes['processedRecords']
+            );
+            //$result = json_encode($notes, JSON_PRETTY_PRINT); // debug
+        }
+        return Okapi::formatted_response($request, $result);
+    }
+
+    // ------------------------------------------------------------------
+
+    private static function parse_notes($field_notes)
+    {
+        $decoded_field_notes = base64_decode($field_notes, true);
+        if ($decoded_field_notes === false) return false;
+
+        $multiline = self::fieldNotesTxtArea2Array($decoded_field_notes);
+        $submittable_logtype_names = Okapi::get_submittable_logtype_names();
+        $records          = [];
+        $totalRecords     = 0;
+        $processedRecords = 0;
+
+        foreach ($multiline as $line) {
+            $totalRecords++;
+            $line = trim($line);
+            $fields = self::CSVtoArray($line);
+
+            $code = $fields[0];
+            $date = $fields[1];
+            $type = $fields[2];
+
+            if (!in_array($type, $submittable_logtype_names)) continue;
+
+            $log  = nl2br($fields[3]);
+
+            $records[] = [
+                'code' => $code,
+                'date' => $date,
+                'type' => $type,
+                'log'  => $log,
+            ];
+            $processedRecords++;
+        }
+        return ['success' => true, 'records' => $records, 'totalRecords' => $totalRecords, 'processedRecords' => $processedRecords];
+    }
+
+
+    // ------------------------------------------------------------------
+
+    private static function fieldNotesTxtArea2Array($fieldnotes)
+    {
+        $output = [];
+        $buffer = '';
+        $start = true;
+
+        $lines = explode("\n", $fieldnotes);
+        $lines = array_filter($lines); // Drop empty lines
+
+        foreach ($lines as $line) {
+            if ($start) {
+                $buffer = $line;
+                $start = false;
+            } else {
+                if (strpos($line, 'OC') !== 0) {
+                    $buffer .= "\n" . $line;
+                } else {
+                    $output[] = trim($buffer);
+                    $buffer = $line;
+                }
+            }
+        }
+
+        if (!$start) {
+            $output[] = trim($buffer);
+        }
+
+        return $output;
+    }
+
+    // ------------------------------------------------------------------
+
+    private static function CSVtoArray($text)
```

Comment:
Of courfse I have tied to use that function and believe me or not, I have spent hours to parse such a string. Ultimately and years ago I wrote those function in JS and they are proven to work so I converted it to php and used it. The problem is in the structure of the CSV record. A CSV record is not a 1:1 match with a "line in a file". The last field of the record, which is a draft log, can be anything the user has composed on the client application. It typicall streches over multiple lines in the fila, it hast control characters in it such as line feeds and carriage returns, it has also single quotes in it, matched or unmached paris, it has double quotes in it, believe me, such trivial functions like `str_getcsv()` miserably fail.

Here is a sampe, however, pls note that the full implications can only be understood if I'd send you a real file. The following file has two records, each 4 fields. The fourth field is tough:

```
OC1012,2023-11-23T10:17:55Z,Found it,"lorem ipsum ipsum lorem

bla
bla

Thx to gcnuf  for 'Brunnen 1: Blick auf Herrenberg von oben',
JoeDante

Take only what you need and be on your way ..."
OC100B,2023-11-23T10:17:46Z,Found it,"Thx to Chaos truppe  for 'cachespätzle',
JoeDante

Take only what you need and be on your way ..."
```

Reply:
Ah, no, there's something off. "OC100B" gets glued to the previous line.

Reply:
Interesting! This looks actually the way it should. I need to investigate, where in tha path the string gets screwed up then, perhaps in the various `encodeURIComponent()` calls. The reason why base64 works is most likely because at no point anything is messing with the string.

More and more I come to the conclusion that we should require the client application to build a clean JSON array which can be stringified at the sending end and parsed at the receiving end.

At this time it would be `cgeo` which would have to adapt. I am not aware of any other client application uploading fieldnotes. Perhaps Garmin can upload directly to geocaching.com, that I don't know for sure but I doubt it.

Reply:
[Here's a fixed version](https://www.tehplayground.com/L0BxcjwFRDm9stmy). I somehow assumed that `str_getcsv` reads the entire CSV, whereas it reads only one row, so this needs to be done in a loop.